File: 3. Embedded Python/src/airco_demo.cpp
#include <iostream>
#include "airco.h"

int main()
{
    std::cout << "Testing Airco class" << std::endl;

    auto callback = [](bool cooler_is_running) {
        std::cout << "[Callback] Airco is" << (cooler_is_running ? "" : " not") << " cooling" << std::endl;
    };

    std::cout << "Creating thermostat" << std::endl;
    auto thermostat = Thermostat();

    std::cout << "Creating controller" << std::endl;
    auto controller = Controller();

    std::cout << "Creating airco" << std::endl;
    auto airco = Airco(controller, thermostat);

    std::cout << "Adding callback function" << std::endl;
    airco.add_callback(callback);

    std::cout << "Trying temperature values (should not give any result)" << std::endl;
    airco.temperature(19.8);
    airco.temperature(20.5);
    airco.temperature(21.2);

    std::cout << "Starting airco" << std::endl;
    airco.start();

    std::cout << "Trying temperature values (should give result)" << std::endl;
    airco.temperature(19.8);
    airco.temperature(20.5);
    airco.temperature(21.2);

    std::cout << "Stopping airco" << std::endl;
    airco.stop();

    std::cout << "Trying temperature values (should not give any result)" << std::endl;
    airco.temperature(19.8);
    airco.temperature(20.5);
    airco.temperature(21.2);

    return 0;
}

File: 3. Embedded Python/src/airco_demo.py
from airco import Airco, Temperature, Thermostat, Controller


if __name__ == '__main__':
    print("Testing airco")

    print("Creating thermostat")
    thermostat = Thermostat()

    print("Creating controller")
    controller = Controller()

    print("Creating airco")
    airco = Airco(controller, thermostat)

    print("Adding callback functions")
    cb = lambda running: print(f"[Callback] airco is{'' if running else ' not'} cooling")
    airco.add_callback(cb)

    print("Starting airco")
    airco.start()
    airco.temperature(20.5)
    airco.temperature(21.2)
    airco.temperature(20.6)
    airco.temperature(19.9)
    airco.temperature(20.4)

    print("Stopping airco")
    airco.stop()

    del airco

    print("Testing inheritance")

    class PyController(Controller):
        def signal(self, event):
            print("[PyController] Received event:", event)
            Controller.signal(self, event)

    print("Creating Python subclass of Controller")
    py_controller = PyController()
    print("Verifying that the derived class works")
    py_controller.signal(Temperature.TEMP_LOW)
    py_controller.signal(Temperature.TEMP_OK)
    py_controller.signal(Temperature.TEMP_HIGH)

    print("Creating airco with inherited controller")
    airco2 = Airco(py_controller, thermostat)
    airco2.add_callback(cb)
    airco2.start()
    print("Setting temperature")
    print("Expected: '[PyController] Received event:'")
    airco2.temperature(20.5)
    airco2.stop()


